### ISSUE TITLE
HW4 - Update

### ISSUE THREAD

**Issue by nevarez-emil** [mikeizbicki/ucr-cs100#1244]
Hi Mike, 

I've got everything up and running finally. View my repo [here](https://github.com/eneva002/rhype).

Let me know if anything breaks!

Happy listening!

**Comment by mikeizbicki** [comment]
Notes:
- I like the brief description you have at the top of the README.  It's
  very descriptive.  I think you should move it up to the main github
  description (where you currently have "A desktop music player for
  hypem.com!").
- You should also add another 2-3 sentences in this first paragraph.  It
  should probably show an example.  Something like:
  
  > To start playing music, just enter the command:
  > 
  > ```
  > $ hypem play
  > ```
  > 
  > To stop playing music ...
- Your title calls your project `rHype` but on github the title is `rhype`
  this was confusing to me when I was trying to install it.  Pick one and be
  consistent.  It doesn't matter which.  They're both good.
- When installing, I get the following error:

```
g++ -std=c++11  -I/usr/include/python2.7 src/rhyped.cpp -o bin/rhyped -ansi
-lpython2.7
src/rhyped.cpp:20:23: fatal error: curl/curl.h: No such file or directory
compilation terminated.
make: *** [rhyped] Error 1
```

I installed all the dependencies you had listed, and I have curl installed,
so I'm not sure what I'm missing.  Maybe it's another python package?
- Your installation instructions need to be more explicit.  Instead of
  saying: "cd into your rhype folder and type `make`", you should say:

> Run the following commands to install:
> 
> ```
> $ git clone https://github.com/eneva002/rhype
> $ cd rhype
> $ make
> $ PATH=$(pwd)/bin
> ```
> 
> It's also customary to add a target called `install` that copies the
> executables into `/usr/local/bin`.  That's not strictly necessary though.

*Under "basic commands", is it intentional that only some of them are
listed as working?  A table is probably a better format for displaying this.

*Under "the future of rhype", you say: "This project is by no means
complete or anywhere near the quality that I'd like it to be".  I like that
you're being honest about the quality, but you don't need to say it in a
negative way.  A better phrasing is: "rhype is still a work in progress.  I
plan to add features X, Y, and Z."

I can give you more detailed feedback about your grade once you get back to
me on the installation issue.

**Comment by nevarez-emil** [comment]
The install should work now, I forgot to delete an included library that I no longer needed.

I'll rework the README right now.

**Comment by nevarez-emil** [comment]
I've also adjusted the README file.

Let me know if there  is anything else that needs revision!

Thanks.

**Comment by mikeizbicki** [comment]
Overall, I'm really impressed!  I think once the rough edges are all polished you'll find a lot of users for this :)

Running make now works, but gives these warnings:

```
g++ -std=c++11  -I/usr/include/python2.7 src/rhyped.cpp -o bin/rhyped -ansi -lpython2.7
In file included from /usr/include/python2.7/Python.h:8:0,
                 from src/rhyped.cpp:20:
/usr/include/python2.7/pyconfig.h:1161:0: warning: "_POSIX_C_SOURCE" redefined [enabled by default]
In file included from /usr/include/x86_64-linux-gnu/sys/socket.h:24:0,
                 from src/rhyped.cpp:1:
/usr/include/features.h:163:0: note: this is the location of the previous definition
In file included from /usr/include/python2.7/Python.h:8:0,
                 from src/rhyped.cpp:20:
/usr/include/python2.7/pyconfig.h:1183:0: warning: "_XOPEN_SOURCE" redefined [enabled by default]
In file included from /usr/include/x86_64-linux-gnu/sys/socket.h:24:0,
                 from src/rhyped.cpp:1:
/usr/include/features.h:165:0: note: this is the location of the previous definition
g++ -std=c++11  src/rhype.cpp -o bin/rhype -ansi -lpython2.7
rm -rf  /home/user/rhype && mkdir /home/user/rhype 
cp src/hyped.py /home/user/rhype/hyped.py
```

People will feel more comfortable using your program if you fix the warnings.

---
- The top of your readme should also mention a bit more about where the songs are coming from.  Assuming it's all legal (I personally don't care either way), you should explicitly say all the music is legal.  You should also say something about what type of music I should expect to listen to when I download it.
- Your README doesn't say anything about `rhyped`, but that's a hugely important part of your program.  Running it should probably be part of the install process.
- When I run `rhype` with no parameters, it should give me a list if possibilities.  Try running `git` by itself to see an example.  This is similar to your help command, but don't make them explicitly run help.  Also, the formatting could be better.
- There's a pretty big delay between when I run `rhype play`/`rhype skip`/etc and when the command actually takes affect.  Do you know why that is?  If it's not easy to fix, this is something that should be listed in the README as a known bug.
- `rhype list` seems to freeze for me.  `rhype pause` did nothing.
- Use the `pwd` command I gave you for updating the PATH. It is guaranteed to always work for all people in all situations without them having to think, and yours requires them to think.
- Your installation procedures involve `rhype play` and other commands.  Why?  This is not part of installation.

**Comment by nevarez-emil** [comment]
The delay is because the file is being downloaded each time. 

If I can figure out a way to get the files playing without storing them locally I will transition to that, but this was the most straightforward approach I could come up with. 

I believe that if I do this in Python I will be able to circumvent the download issue. I'm not 100% on that, but I think it is possible.

I'm not sure about your issue with `list`, or the wall of warnings, but I'll look into them. `pause`, `stop`, and `skip` are all built using `signal()`, so I think that if I transition to sigaction that may fix the issue.

I've updated the README, let me know what you think.

**Comment by mikeizbicki** [comment]
The first line in your README needs to be a complete sentence.  It should also not be standing alone like that; merge it with the next paragraph.

`The most popular artists, searches and blogs on the internet right now` shouldn't be in blockquotes.  Just use double quotes for this.

You're intro needs to say something about how easy it is to use because you just have to type `rhype play` to get music.

Otherwise, your writing and tone is great throughout the README.

How difficult would it be to change "stations" on hypem from popular to something else?  I'm guessing it wouldn't be too hard if all the pages have similar formatting.  That'd be a really nice feature.  I'd suggest you add that before publicizing your project.

Dependencies should be a subsection under Installing.  

Running `rhyped` isn't really part of the installation procedure.  So I would put that in a separate code block saying something like:

> Once you've completed these steps, just run the rhype daemon:
> 
> ```
> $ rhyped
> ```
> 
> And you're ready to start listening to music!

Change "Working details" to "Implementation details"

You wrote: "All working files for rhype will be stored ...", but you should say "are" instead of "will be".  Always use the present tense in documentation.

Your two error messages should be on separate lines within a numbered list (you'll have to lookup the markdown formatting for this).

---

> I could download them all at once, but I think that would open up too many potential legal issues.

If it's okay to download one video, it'll be okay to download whatever the next 1-2 videos are so that they're preloaded.  You don't have to download the whole library.  If you do have to wait to download something, then you should print out a message so the user knows something is going on.

Also, just FYI, you should never express in writing (esp. on a public forum) that you have legal concerns about a project you're working on.  That can only come back to haunt your project later.  Lawyers can point to your statements as evidence that you knowingly did something illegal.  That said, I'm pretty confident you're okay, and it's almost always fine to just do things and then let the lawyers stop you if they find it objectional.  If you have any questions about this, email me.

---

I'm giving you a 120/100 under the expectation that you'll fix all the issues above.  I recommend once you fix them, let me review it one last time before starting to publicize it.  You've done a great job with this project!  You should be proud of it :)
